### PR TITLE
Am/m3/pid2

### PR DESCRIPTION
**Pull request by csabahi**
## Description

*What was completed, changed, or updated?* 

NOT ready to merge yet, waiting for IMU filters and driver to be completed
Designed PIDs for ZP3.5
Implemented it in Manual and FBWA flightmodes

<br/>

*Why was this done (if applicable)?*

Because WARG need PID

<br/>

## Testing

*What **manual** tests were used to validate the code?*

<br/>

*What **unit** tests were used to validate the code?*

Wrote unit tests in PIDTests.cpp

<br/>

## Documentation

*Milestone number and name:*
M3
*Link to Asana task:*
https://app.asana.com/0/1203458353737758/1206541298938863/f

*Link to Confluence documentation:*
https://uwarg-docs.atlassian.net/wiki/x/AQCtjw
<br/>

## Reminders

- [ ] Add reviewers to the PR

- [ ] Mention the PR in the appropriate discord channel

### CODE REVIEW

File: AttitudeManager/FlightModes/Inc/flightmode.hpp [hunk 1]
```diff
@@ -10,15 +10,17 @@
...
-
+#include <PID.hpp>
 namespace AM {
 
 class Flightmode {
    public:
     virtual ~Flightmode() = default;
 
-    virtual AttitudeManagerInput run(const AttitudeManagerInput& input) = 0;
+    virtual AttitudeManagerInput run(const AttitudeManagerInput& input, AxisPIDs) = 0;
```
Comment (antholuo) [inline AttitudeManager/FlightModes/Inc/flightmode.hpp:20]: we shouldn't be passing AxisPIDs in as a function to our AM ::run function....

We should be able to grab the values from private class variables.

Those PID's should only be set using the updatePIDGains or in the initialization function.

File: AttitudeManager/FlightModes/Inc/fbwa.hpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+#include "flightmode.hpp"
+
+namespace AM {
+
+class FBWA : public Flightmode {
+   public:
+    FBWA() = default;
+
+    AttitudeManagerInput run(const AttitudeManagerInput& input, AxisPIDs _pids) override;
```
Comment (antholuo) [inline AttitudeManager/FlightModes/Inc/fbwa.hpp:12]: our PID's should exist as private class variables, and not have to be passed into the "run" function.

They can be set during the initialization of the class! (in the constructor).

File: AttitudeManager/FlightModes/Src/fbwa.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            fbwa_control_limits.throttleLimit.min +
+            (input.throttle - AM::AttitudeManager::INPUT_MIN) *
+            (fbwa_control_limits.throttleLimit.max - fbwa_control_limits.throttleLimit.min) /
+            (AM::AttitudeManager::INPUT_MAX - AM::AttitudeManager::INPUT_MIN);
+        
+        //Need to get actual and actualRate data from IMU filter
+        // _pids.pitch.execute(mappedOutputs.pitch, actual, actualRate);
+        // _pids.roll.execute(mappedOutputs.roll, actual, actualRate);
+        // _pids.yaw.execute(mappedOutputs.yaw, actual, actualRate);   
+        // _pids.throttle.execute(mappedOutputs.throttle, actual, actualRate);
```
Comment (antholuo) [inline AttitudeManager/FlightModes/Src/fbwa.cpp:44]: no pid needed on throttle!

File: AttitudeManager/FlightModes/Src/fbwa.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        mappedOutputs.throttle =
+            fbwa_control_limits.throttleLimit.min +
+            (input.throttle - AM::AttitudeManager::INPUT_MIN) *
+            (fbwa_control_limits.throttleLimit.max - fbwa_control_limits.throttleLimit.min) /
+            (AM::AttitudeManager::INPUT_MAX - AM::AttitudeManager::INPUT_MIN);
+        
+        //Need to get actual and actualRate data from IMU filter
+        // _pids.pitch.execute(mappedOutputs.pitch, actual, actualRate);
+        // _pids.roll.execute(mappedOutputs.roll, actual, actualRate);
+        // _pids.yaw.execute(mappedOutputs.yaw, actual, actualRate);
```
Comment (antholuo) [inline AttitudeManager/FlightModes/Src/fbwa.cpp:43]: yaw is unique, it operates on yawrate! So you'll need to give it your rate, and then acceleration!

File: AttitudeManager/FlightModes/Src/manual.cpp [hunk 5]
```diff
@@ -3,10 +3,28 @@
...
+    switch (whichGain) {
+        case proportional:
+            _axis.setKp(desiredGain);
+            break;
+        case integral:
+            _axis.setKi(desiredGain);
+            break;
+        case derivative:
+            _axis.setKd(desiredGain);
+            break;
```
Comment (antholuo) [inline AttitudeManager/FlightModes/Src/manual.cpp:25]: Remember to have a default case which throws some sort of error/ warning!

Also remember to do bounds checking on your whichGain :)

File: AttitudeManager/FlightModes/Src/manual.cpp [hunk 6]
```diff
@@ -3,10 +3,28 @@
 
 namespace AM {
 
-AttitudeManagerInput Manual::run(const AttitudeManagerInput& input) {
+AttitudeManagerInput Manual::run(const AttitudeManagerInput& input, AxisPIDs _pids) {
     return input;
 }
 
 void Manual::updatePid() {}
+void Manual::updatePidGains(PIDController _axis, GainTerm whichGain, float desiredGain) {
```
Comment (antholuo) [inline AttitudeManager/FlightModes/Src/manual.cpp:11]: You shouldn't need to pass in a PIDController class - you should be able to pass in the "axis" that you would like to tune ('whichAxis', for example), and your function should be able to grab the flightmodes private PID class.

This follows the "no external access to the PID's". Each class owns it's own PID controllers.

Also - "whichGain" or "whichAxis" probably aren't my top picks for variable names, I'd rather `pid_axis` and `pid_gain_term`.

I'm not sure what our code style says for snake vs camel case, but whatever it says goes good w/me.

File: AttitudeManager/FlightModes/Src/fbwa.cpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        // _pids.throttle.execute(mappedOutputs.throttle, actual, actualRate);
+
+
+        return mappedOutputs;
+
+    }
+
+    void FBWA::updatePid() {} //Needs to be implemented
+
+    void FBWA::updatePidGains(PIDController _axis, GainTerm whichGain, float desiredGain) {
```
Comment (antholuo) [inline AttitudeManager/FlightModes/Src/fbwa.cpp:53]: I didn't see this line for some reason, but comments I made on the manual () class apply here as well.

File: AttitudeManager/Src/AM.cpp [hunk 8]
```diff
@@ -19,6 +19,19 @@ AttitudeManagerInput AttitudeManager::control_inputs = {
     .throttle = 0.0f
 };
 
+AxisPIDs axisPIDs = {
```
Comment (antholuo) [inline AttitudeManager/Src/AM.cpp:22]: You should be able to declare these in the AM::fbwa class (same ideas as prev. comments)

You shouldn't have them in the main AM.cpp at all, since there will be more flight modes in the future which will also have varying different PID's.

File: AttitudeManager/Src/AM.cpp [hunk 9]
```diff
@@ -49,7 +62,7 @@ void AttitudeManager::runControlLoopIteration() {
     AttitudeManagerInput control_inputs = getControlInputs();
 
     // Run Control Algorithms
-    AttitudeManagerInput motor_outputs = controlAlgorithm_->run(control_inputs);
+    AttitudeManagerInput motor_outputs = controlAlgorithm_->run(control_inputs, axisPIDs);
```
Comment (antholuo) [inline AttitudeManager/Src/AM.cpp:65]: above comment(s) - PID owned by class and we shouldn't need to pass them. 

TLDR: the function that calls `<flightmode>::run` doesn't know or care what PID's exist and how they need to be managed. It just expects the flightmode to update and give motor outputs when `run` is called.

File: AttitudeManager/Src/PID.cpp [hunk 10]
```diff
@@ -11,28 +11,73 @@
...
-        // first order derivative computation
-        derivative = ((3 * historicalValue[0]) - (4 * historicalValue[1]) + (historicalValue[2]));
+        derivative = execute_d_back(_desired, _actual, _actualRate);
+        //One is uncommented to test diff between types of calculation
+        // derivative = execute_d_hist(_desired, _actual, _actualRate);
     }
 
-    float ret = constrain<float>((pid.kp * error) + (pid.ki * integral) - (pid.kd * derivative),
-                                 pid.max_output, pid.min_output);
+    float ret = constrain<float>((pid.kp * error) + (pid.ki * integral) - (derivative),
```
Comment (antholuo) [inline AttitudeManager/Src/PID.cpp:38]: where did `pid.kd` go ?

File: AttitudeManager/Tests/PIDTest.cpp [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    ASSERT_TRUE(pidController.kp == newGains.kp);
+    ASSERT_TRUE(pidController.ki == newGains.ki);
+    ASSERT_TRUE(pidController.kd == newGains.kd);
+}
+
+
+int main(int argc, char** argv) {
+    ::testing::InitGoogleTest(&argc, argv);
+    return RUN_ALL_TESTS();
+}
```
Comment (antholuo) [inline AttitudeManager/Tests/PIDTest.cpp:150]: newline @ eof
```suggestion
}

```

File: AttitudeManager/Src/PID.cpp [hunk 12]
```diff
@@ -11,28 +11,73 @@
...
+float PIDController::execute_d_back(float _desired, float _actual, float _actualRate){
+    float error = _desired - _actual;
+    float derivative = error - prevError;
+    prevError = error;
+    return pid.kd * derivative;
+}
+
+float PIDController::map(float x, float in_min, float in_max, float out_min, float out_max) {
+    return (x - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+}
```
Comment (antholuo) [inline AttitudeManager/Src/PID.cpp:83]: newline @ eof
```suggestion
}

```

**Review by antholuo — CHANGES_REQUESTED** [review]
Looks like it needs a good rebase, and is waiting on some adjacent structs.

I had a few things that I may not have explained super clearly, but I tried my best in the comments. Let me know and just ping me if you're confused about what I said.

I also noticed some weird formatting. I'd suggest running a clang format using warg's template or your own preffered template. It'll get rid of some weird whitespace / newline shenanigans that seem to be popping up here and there.

Overall, this looks super awesome!!
